dbms/src/DataStreams/WindowBlockInputStream.cpp

```diff
@@ -28,6 +40,232 @@ extern const int BAD_ARGUMENTS;
+
+        if constexpr (checkIfDecimalFieldType<LeftType>())
+        {
+            left_scale = left.getScale();
+            return DecimalComparison<typename LeftType::DecimalType, RightType, LessOrEqualsOp>::compare(
+                left.getValue(),
+                right,
+                left_scale,
+                right_scale);
+        }
+        else if constexpr (checkIfDecimalFieldType<RightType>())
+        {
+            right_scale = right.getScale();
+            return DecimalComparison<LeftType, typename RightType::DecimalType, LessOrEqualsOp>::compare(
+                left,
+                right.getValue(),
+                left_scale,
+                right_scale);
+        }
+    }
+    else
+    {
+        return left <= right;
+    }
+}
+
+template <typename LeftType, typename RightType>
+bool greaterEqual(LeftType left, RightType right)
+{
+    if constexpr (checkIfDecimalFieldType<LeftType>() && checkIfDecimalFieldType<RightType>())
+    {
+        // This `if` branch is necessary or the next branch's codes will be generated
+        // when these two types are all DecimalField type which will cause compilation errors.
+        return left >= right;
+    }
+    else if constexpr (checkIfDecimalFieldType<LeftType>() || checkIfDecimalFieldType<RightType>())
+    {
+        // Only one type will be Decimal as Decimal vs Decimal can be handled in the `else` branch
+        UInt32 left_scale = 0;
+        UInt32 right_scale = 0;
+
+        if constexpr (checkIfDecimalFieldType<LeftType>())
+        {
+            left_scale = left.getScale();
+            return DecimalComparison<typename LeftType::DecimalType, RightType, GreaterOrEqualsOp>::compare(
+                left.getValue(),
+                right,
+                left_scale,
+                right_scale);
+        }
+        else if constexpr (checkIfDecimalFieldType<RightType>())
+        {
+            right_scale = right.getScale();
+            return DecimalComparison<LeftType, typename RightType::DecimalType, GreaterOrEqualsOp>::compare(
+                left,
+                right.getValue(),
+                left_scale,
+                right_scale);
+        }
+    }
+    else
+    {
+        return left >= right;
+    }
+}
+
+// When T is Decimal, we should convert it to DecimalField type
+// as we need scale value when executing the comparison operation.
+template <typename T>
+struct ActualCmpDataType
+{
+    using Type = std::conditional_t<checkIfSimpleNumericType<T>(), T, DecimalField<T>>;
+};
+
+template <typename T>
+typename ActualCmpDataType<T>::Type getValue(const ColumnPtr & col_ptr, size_t idx)
+{
+    return (*col_ptr)[idx].get<typename ActualCmpDataType<T>::Type>();
+}
+
+template <typename T, typename U, bool is_preceding, bool is_desc, bool is_begin>
+bool isInRangeCommonImpl(T current_row_aux_value, U cursor_value)
+{
+    if constexpr (is_begin)
+    {
+        if constexpr ((is_preceding && is_desc) || (!is_preceding && is_desc))
```

Comment:
I don't get why `is_preceding` is needed here? Looks like `((is_preceding && is_desc) || (!is_preceding && is_desc))` is the same as `is_desc`?

Reply:
> I don't get why `is_preceding` is needed here? Looks like `((is_preceding && is_desc) || (!is_preceding && is_desc))` is the same as `is_desc`?

From a logical perspective, which if-branch we choose to execute is decided by `is_preceding` and `is_desc`, so it's easier to understand the purpose of this if-condition if we use both `is_preceding` and `is_desc`

Reply:
changed